src/Data/HashTable/ST/Basic.hs

```diff
@@ -460,56 +483,46 @@ growTable ht@(HashTable sz _ _ _ _) = do
 
 
 ------------------------------------------------------------------------------
--- Helper data structure for delete'
+-- Helper data structure for findSafeSlots
 data Slot = Slot {
       _slot       :: {-# UNPACK #-} !Int
-    , _wasDeleted :: {-# UNPACK #-} !Int  -- we use Int because Bool won't
-                                          -- unpack
     }
   deriving (Show)
 
 
 ------------------------------------------------------------------------------
 instance Monoid Slot where
-    mempty = Slot maxBound 0
-    (Slot x1 b1) `mappend` (Slot x2 b2) =
-        if x1 == maxBound then Slot x2 b2 else Slot x1 b1
+    mempty = Slot maxBound
+    (Slot x1) `mappend` (Slot x2) =
+        if x1 == maxBound then Slot x2 else Slot x1
 
 
 ------------------------------------------------------------------------------
--- Returns the slot in the array where it would be safe to write the given key.
-delete' :: (Hashable k, Eq k) =>
-           (HashTable_ s k v)
-        -> Bool
-        -> k
-        -> Int
-        -> ST s Int
-delete' (HashTable sz loadRef hashes keys values) clearOut k h = do
-    debug $ "delete': h=" ++ show h ++ " he=" ++ show he
+-- Returns a touple ST s (found, b0, b1) where
+--     * found :: Bool - True if key-value mapping is already in the table.
+--     * b0 :: Int     - The index of a slot where it would be safe to write
+--                       the given key (if the key is already in the mapping,
+--                       you have to delete it before using this slot).
+--     * b1 :: Int     - The index of a slot where the key currently resides.
+--                       Or, if the key is not in the table, b1 is a slot
+--                       where it is safe to write the key (b1 == b0).
+findSafeSlots :: (Hashable k, Eq k) =>
+                 (HashTable_ s k v)
+              -> k
+              -> Int
+              -> ST s (Bool, Int, Int)
```

Comment:
I think it might we worth defining a strict tuple here to save on allocation:

```haskell
data SlotFindResponse = SlotFindResponse {
    _slotFound :: {-# UNPACK #-} !Bool
  , _slotB0 :: {-# UNPACK #-} !Int
  , _slotB1 :: {-# UNPACK #-} !Int
}
```